### ISSUE TITLE
[Hold #36071] [$500] Chat - Markdown is pasted only once when message with markdown copied

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#36269]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.39-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4292604
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Log in to New Expensify
2. Navigate to any chat
3. Send message with markdown and copy it
4. Paste copied message in compose box at least twice

## Expected Result:
Everytime the message is pasted, it should contain copied markdown

## Actual Result:
Message's markdown is only pasted once. Second time and onwards message is pasted as plain text

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details>  

https://github.com/Expensify/App/assets/78819774/4368652f-5be3-428b-9d05-1dbb37edb1bd

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~015f143311cf0c364a</li>
        <li>Upwork Job ID: 1755995389655240704</li>
        <li>Last Price Increase: 2024-02-09</li>
    </ul>
</details>

**Comment by lanitochka17** [comment]
We think that this bug might be related to #vip-vsp
CC @quinthar

**Comment by ikevin127** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue

Message's markdown is only parsed once if pasted one right after the other, the second time and onwards the message is not being parsed, instead remains as plain text.

### What is the root cause of that problem?

The issue comes from our [expensify-common/ExpensiMark.js](https://github.com/Expensify/expensify-common/blob/main/lib/ExpensiMark.js) library which we use to parse the chat input text to markdown.

The problem lies in the incorrect usage of word boundaries (`\b` and `\B`) in regex patterns, which do not correctly match markdown syntax across sequences.

### What changes do you think we should make in order to solve the problem?

A fix for bold and italic only would look something like this:

```javascript
{
    name: 'italic',
    regex: /(?:^|\s|>)(_)(?=\S)([\s\S]*?\S)_(?=$|\s|<)/g,
    replacement: '$1<em>$2</em>$3',
},

{
    name: 'bold',
    regex: /(?:^|\s|>)(\*)(?=\S)([\s\S]*?\S)\*(?=$|\s|<)/g,
    replacement: '$1<strong>$2</strong>$3',
},
```

This modification adds a non-capturing group `(?:^|\s|>)` to check for a start of string, a whitespace, or a `>` character before the markdown symbol and `(?=$|\s|<)` to check for an end of string, a whitespace, or a `<` character after the markdown symbol. These changes help ensure that markdown styles are correctly applied throughout the text.

**Comment by ikevin127** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue

Message's markdown is only parsed once if pasted one right after the other, the second time and onwards the message is not being parsed, instead remains as plain text.

### What is the root cause of that problem?

The copy to clipboard uses this function to get the text only version of the message:

https://github.com/Expensify/App/blob/c3555eddaad725625c86445b355647ccb09ca61c/src/pages/home/report/ContextMenu/ContextMenuActions.tsx#L39-L42

and since this only extracts the text, when pasted multiple times back in the composer this will be plain text instead of markdown.

### What changes do you think we should make in order to solve the problem?

Change the copy to clipboard `getActionText` function to check if the `curr?.text === curr?.html` and if they are **not equal** this means the message contains markdown.

In this case we need to use the ExpensiMark library and return `parser.replace('${acc}${curr.html}')` which will parse the html message into markdown then copy it to the clipboard instead of copying the plain text message.

**Comment by thienlnam** [comment]
This bug no longer exists, adding it to the tracking issue

**Comment by trjExpensify** [comment]
@thienlnam would you like to close this?

**Comment by thienlnam** [comment]
I'm holding onto them so we can track that they're fixed in the new PR and don't get recreated once we merge again

**Comment by trjExpensify** [comment]
Okay, I've made that clear in the title.

**Comment by thienlnam** [comment]
Same as above

**Comment by BartoszGrajdek** [comment]
@thienlnam we should be able to close this one now
